transient/config.py

```diff
@@ -0,0 +1,261 @@
+#
+# The only difference is that the QEMU argument string is split by whitespace
+# and encapsulated in a list.
+#
+
+[qemu]
+
+qemu-args = [
+    "-nographic",
+    "-enable-kvm",
+    "-m", "1G"
+]
+"""
+
+
+class Config(Dict[Any, Any]):
+    """Creates an argument dictionary that allows dot notation to access values
+
+    Example:
+
+        >>> args = Config({'arg1': 1, 'arg2': 2})
+        >>> args['arg1'] == args.arg1
+
+    """
+
+    def __init__(self, *args: Dict[Any, Any], **kwargs: Dict[Any, Any]) -> None:
+        super(Config, self).__init__(*args, **kwargs)
+        for arg in args:
+            if isinstance(arg, dict):
+                for k, v in arg.items():
+                    self[k] = v
+
+        if kwargs:
+            for k, v in kwargs.items():
+                self[k] = v
+
+    def __getattr__(self, attr: Any) -> Any:
+        return self.get(attr)
+
+    def __setattr__(self, key: Any, value: Any) -> None:
+        self.__setitem__(key, value)
+
+    def __setitem__(self, key: Any, value: Any) -> None:
+        super(Config, self).__setitem__(key, value)
+        self.__dict__.update({key: value})
+
+    def __delattr__(self, item: Any) -> None:
+        self.__delitem__(item)
+
+    def __delitem__(self, key: Any) -> None:
+        super(Config, self).__delitem__(key)
+        del self.__dict__[key]
+
+
+class TransientConfigSchema(Schema):
+    """Defines the schema for Transient-run arguments and validates the fields
+       during deserialization
+    """
+
+    config = fields.Str(allow_none=True)
+    copy_in_before = fields.List(fields.Str(), allow_none=True)
+    copy_out_after = fields.List(fields.Str(), allow_none=True)
+    copy_timeout = fields.Int(allow_none=True)
+    image = fields.List(fields.Str(), allow_none=True)
+    image_backend = fields.Str(allow_none=True)
+    image_frontend = fields.Str(allow_none=True)
+    name = fields.Str(allow_none=True)
+    prepare_only = fields.Bool(allow_none=True)
+    qemu_args = fields.List(fields.Str(), allow_none=True)
+    qmp_timeout = fields.Int(allow_none=True)
+    shutdown_timeout = fields.Int(allow_none=True)
+    ssh_command = fields.Str(allow_none=True)
+    ssh_bin_name = fields.Str(allow_none=True)
+    ssh_port = fields.Int(allow_none=True)
+    ssh_timeout = fields.Int(allow_none=True)
+    ssh_user = fields.Str(allow_none=True)
+    ssh_console = fields.Bool(allow_none=True)
+    ssh_with_serial = fields.Bool(allow_none=True)
+    shared_folder = fields.List(fields.Str(), allow_none=True)
+
+    # marshmallow's decorator post_load() is untyped, forcing create_args()
+    # to be untyped. Therefore, we ignore it to silence the type checker
+    @post_load  # type: ignore
+    def create_config(self, data: Dict[Any, Any], **kwargs: Dict[Any, Any]) -> Config:
+        """Returns the Config dictionary after a schema is loaded and validated
+        """
+        return Config(**data)
+
+
+def replace_hyphens_with_underscores_in_dict_keys(
```

Comment:
As we aren't explicitly listing exports, this should start with a `_` to indicate it is local to this module.

Reply:
Prefix added.